Date: Dec 13, 2019 4:05 PM
From: Change-Id: I56605d17f33c04ea06658184e0d5cf1f00280188
Subject: RDKB-22373: Upgrade lighttpd to 1.4.53
Porting the patches added to lighttpd-1.4.45 version to latest 1.4.53 version as well.
Older patches are slightly modified to make it compatible with latest version.
Context refreshed for lighttpd 1.4.74: r -> rc, EAI_ADDRFAMILY guard now also
tests __GLIBC__, log_error_write() -> log_serror(), and set_v6only is tri-state.
The inserted SO_BINDTODEVICE blocks are unchanged from the 1.4.53 original.
Source: COMCAST
Upstream-Status: Pending
Signed-off-by: Rakhil P E <Rakhil_PuthiyaveettilEdachena@comcast.com>

--- a/src/sock_addr.c	2026-09-09 11:57:11.967378717 +0000
+++ b/src/sock_addr.c	2026-09-09 11:57:11.971378738 +0000
@@ -494,6 +494,9 @@
             hints.ai_protocol = IPPROTO_TCP;
 
             if (0 != (rc = getaddrinfo(str, NULL, &hints, &res))) {
+            #ifdef SO_BINDTODEVICE
+                saddr->ipv6.sin6_addr = in6addr_any;
+            #else
                 hints.ai_family = AF_INET;
                 if (
                   #if defined(__GLIBC__) && defined(EAI_ADDRFAMILY)
@@ -513,10 +516,13 @@
                   "getaddrinfo failed: %s '%s'", gai_strerror(rc), str);
 
                 return 0;
+            #endif
             }
 
+            #ifndef SO_BINDTODEVICE
             memcpy(saddr, res->ai_addr, res->ai_addrlen);
             freeaddrinfo(res);
+            #endif
         }
         saddr->ipv6.sin6_port = htons(port);
         *len = sizeof(struct sockaddr_in6);
@@ -532,6 +538,9 @@
             saddr->ipv4.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
         }
         else {
+#ifdef SO_BINDTODEVICE
+          saddr->ipv4.sin_addr.s_addr = htonl(INADDR_ANY);
+#else
           #ifdef HAVE_INET_PTON
             /*(reuse HAVE_INET_PTON for presence of getaddrinfo())*/
             struct addrinfo hints, *res;
@@ -572,6 +581,7 @@
             memcpy(&saddr->ipv4.sin_addr.s_addr,
                    he->h_addr_list[0], he->h_length);
           #endif
+#endif
         }
         saddr->ipv4.sin_port = htons(port);
         *len = sizeof(struct sockaddr_in);
--- a/src/network.c	2026-09-09 11:57:11.967378717 +0000
+++ b/src/network.c	2026-09-09 11:57:11.971378738 +0000
@@ -556,11 +556,13 @@
 
 #ifdef HAVE_IPV6
 		if (set_v6only) {
+#ifndef SO_BINDTODEVICE
 				int val = (set_v6only > 0);
 				if (-1 == setsockopt(srv_socket->fd, IPPROTO_IPV6, IPV6_V6ONLY, &val, sizeof(val))) {
 					log_serror(srv->errh, __FILE__, __LINE__, "setsockopt(IPV6_V6ONLY)");
 					return -1;
 				}
+#endif
 		}
 #endif
 	}
